### PR TITLE
First pass at introspection endpoint

### PR DESCRIPTION
**Pull request by TimothyBJacobs**
Trac ticket: https://core.trac.wordpress.org/ticket/52275

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: tests/phpunit/tests/rest-api/rest-application-passwords-controller.php [hunk 1]
```diff
@@ -69,6 +69,12 @@ public function setUp() {
 		add_filter( 'wp_is_application_passwords_available', '__return_true' );
 	}
 
+	public function tearDown() {
+		parent::tearDown();
+
+		unset( $_SERVER['PHP_AUTH_USER'], $_SERVER['PHP_AUTH_PW'], $GLOBALS['wp_rest_application_password_status'], $GLOBALS['wp_rest_application_password_uuid'] );
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/rest-api/rest-application-passwords-controller.php:75]:
```suggestion
		unset( $_SERVER['PHP_AUTH_USER'], $_SERVER['PHP_AUTH_PW'], $GLOBALS['wp_rest_application_password_status'], $GLOBALS['wp_rest_application_password_uuid'] );
		parent::tearDown();
```

Nit: This matches how it's done in _most_ tests, but it's a bit of a mix within core.

File: src/wp-includes/rest-api.php [hunk 2]
```diff
@@ -1050,13 +1050,36 @@ function rest_cookie_collect_status() {
...
+function rest_application_password_collect_status( $user_or_error, array $app_password = array() ) {
+	global $wp_rest_application_password_status, $wp_rest_application_password_uuid;
 
 	$wp_rest_application_password_status = $user_or_error;
+
+	if ( $app_password ) {
+		$wp_rest_application_password_uuid = $app_password['uuid'];
+	} else {
+		$wp_rest_application_password_uuid = null;
+	}
```
Comment (peterwilsoncc) [inline src/wp-includes/rest-api.php:1067]: Logic check that `uuid` is set in the array.
```suggestion
	if ( empty( $app_password['uuid'] ) ) {
		$wp_rest_application_password_uuid = null;
	} else {
		$wp_rest_application_password_uuid = $app_password['uuid'];
	}
```

File: src/wp-includes/rest-api/endpoints/class-wp-rest-application-passwords-controller.php [hunk 3]
```diff
@@ -373,6 +389,70 @@ public function delete_item( $request ) {
...
+				__( 'The authenticated Application Password can only be introspected for the current user.' ),
+				array( 'status' => rest_authorization_required_code() )
+			);
+		}
+
+		return true;
+	}
+
+	/**
+	 * Retrieves the application password being currently used for authentication.
```
Comment (peterwilsoncc) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-application-passwords-controller.php:419]: I'm trying to work the purpose of this function out, shouldn't the app know the password already by its ability to reach this point?

**Review by peterwilsoncc — COMMENTED** [review]
Added a few notes but not sure I am the best to review as this isn't something I've p[aid a huge amount of attention to, sorry.

File: tests/phpunit/tests/auth.php [hunk 4]
```diff
@@ -44,7 +44,7 @@ public function tearDown() {
 		parent::tearDown();
 
 		// Cleanup all the global state.
-		unset( $_SERVER['PHP_AUTH_USER'], $_SERVER['PHP_AUTH_PW'], $GLOBALS['wp_rest_application_password_status'] );
+		unset( $_SERVER['PHP_AUTH_USER'], $_SERVER['PHP_AUTH_PW'], $GLOBALS['wp_rest_application_password_status'], $GLOBALS['wp_rest_application_password_uuid'] );
```
[the remark this answers is not in the document]
Comment (TimothyBJacobs) [reply tests/phpunit/tests/auth.php:47]: That's probably safer, I'll adjust.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-application-passwords-controller.php (hunk 3 above)
Comment (TimothyBJacobs) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-application-passwords-controller.php:419]: The app knows its password, but not its `uuid`. The `uuid` is how they are queried, but the password is how they are authenticated.

File: tests/phpunit/tests/auth.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (TimothyBJacobs) [reply tests/phpunit/tests/auth.php:49]: So I gave this a go, but I couldn't figure out when these would be pre-filled to a value that should be restored and it added a fair bit of complexity to properly handle.

**Comment by TimothyBJacobs** [comment]
Merged in https://github.com/WordPress/wordpress-develop/commit/e290a9b557406230612161839d90b068f3c6983b.

**Pull request closed without merging**